cupy/functional/piecewise.py

```diff
@@ -0,0 +1,73 @@
+import cupy
+
+from cupy import core
+
+_piecewise_krnl = core.ElementwiseKernel(
+    'U condlist, raw S funclist, int64 xsiz',
+    'raw I prev, raw T y',
+    '''
+        if (condlist && prev[i % xsiz] < i)
+            prev[i % xsiz] = i;
+        __syncthreads();
+        if (prev[i % xsiz] == i)
+            y[i % xsiz] = funclist[i / xsiz];
+        ''',
+    'piecewise_kernel'
+)
+
+
+def piecewise(x, condlist, funclist):
+    """Evaluate a piecewise-defined function.
+
+        Args:
+            x: input domain
+            condlist: conditions list ( boolean arrays or boolean scalars).
+                Each boolean array/ scalar corresponds to a function
+                in funclist. Length of functionlist is equal to that
+                of condlist. If one extra function is given, it is used
+                as otherwise condition
+            funclist: list of scalar functions.
+
+        Returns:
+            cupy.ndarray: the result of calling the functions in funclist
+                on portions of x defined by condlist.
+
+        .. warning::
+
+            This function currently doesn't support callable functions
+            args and kw parameters are not supported
+
+        .. seealso:: :func:`numpy.piecewise`
+        """
+
+    if any(callable(item) for item in funclist):
+        raise NotImplementedError(
+            'Callable functions are not supported currently')
+    if cupy.isscalar(x):
+        x = cupy.asarray(x)
```

Comment:
We don't need to support scalar inputs because it will not work with good performance.

Reply:
Should I raise error or not handle it at all (and eliminate its corresponding test) ?